Mrc 5146 container gh actions

M-Kusumgar opened the pull request:
This turned into a longer PR than I had initially hoped for. This PR builds docker and pushes it dockerhub and github container registry. Main changes include:
* workflow for pushing
* upgrading R version in the package because R versions less than 4.1.0 have problems with decoding url params
* added host option in server main entrypoint
* added worker main entrypoint that also creates their worker directory in `.packit`
* `REDIS_CONTAINER_NAME` env var so that we can connect to redis running in another container
* various scripts for testing this out fully in a docker setup, these are outlined in the README

M-Kusumgar committed:
add entrypoints + build and push gh actions workflow

M-Kusumgar committed:
actually reference dockerfile

M-Kusumgar committed:
install rrq, orderly2 and porcelain

M-Kusumgar committed:
add run on push to this branch

M-Kusumgar committed:
try install hiredis too

M-Kusumgar committed:
install libhiredis-dev

M-Kusumgar committed:
clean up + login to ghcr

M-Kusumgar committed:
chmod entrypoint scripts

M-Kusumgar committed:
install git in dockerfile

M-Kusumgar committed:
install redis for debugging on image

M-Kusumgar committed:
redis name via env var for containers

M-Kusumgar committed:
use worker loop to keep docker container alive

M-Kusumgar committed:
remove unnecessary scripts and setup worker dir

M-Kusumgar committed:
include test with containers

M-Kusumgar committed:
better clear script

M-Kusumgar committed:
have a minimum version for R and update docker build version

M-Kusumgar committed:
update description file to contain stringr

M-Kusumgar committed:
try r 4.2

M-Kusumgar committed:
try r 4.1

M-Kusumgar committed:
use volume for orderly root + git config

M-Kusumgar committed:
more scripts

M-Kusumgar committed:
tests

M-Kusumgar committed:
add readme section on testing

M-Kusumgar commented on docker/Dockerfile:
```diff
@@ -17,3 +19,15 @@ RUN install2.r --error remotes &&  \
+
+COPY docker/bin /usr/local/bin/
+
+RUN git config --global --add safe.directory /orderly-root
+RUN echo ".packit" > /.gitignore
+RUN git config --global core.excludesFile "/.gitignore"
```
same issue as in outpack_server PR, gotta add this as a safe directory as it is not owned by us in the container, also the global gitignore means that the .packit where the workers are isnt tracked by git

absternator commented on docker/test/run-test:
```diff
@@ -0,0 +1,41 @@
+
+docker volume create orderly-root-volume
+
+docker run --rm -d --pull=missing \
```
Q: whats this one for?

M-Kusumgar replied:
that one is for a couple things:
1. you need a volume attached to a container to do anything with it so for us modifying the orderly repo, we could use this as the place to change things in the volume for testing
2. to view how things change in the working directory (since workers will be running things and producing result files and metadata of the run) theres a helper script `docker/test/copy-orderly-root` that will copy the folder so we can see it easily - that script uses this container too

absternator commented on docker/test/run-test:
```diff
@@ -0,0 +1,41 @@
+
+docker network create orderly.runner_network
+
+docker run --rm -d \
```
also wondering if redis is needed here? or maybe another script that dosent... as this will all be brought up with packet (which already brings up redis), thus will get conflict

richfitz replied:
we do need redis here, but that's fine for testing this stuff. You will get a conflict if two containers bring up redis on the same port, but in practice it's not hard to deal with

absternator commented on docker/Dockerfile:
```diff
@@ -17,3 +19,15 @@ RUN install2.r --error remotes &&  \
 RUN R CMD INSTALL --install-tests /src && rm -rf /src
+
+ENV ORDERLY_RUNNER_QUEUE_ID=orderly.runner
+ENV REDIS_CONTAINER_NAME=orderly.runner_redis
```
will this come from packet? or will this have another redis container?

M-Kusumgar replied:
packit does not need a redis container, after addressing rich's comments, this will come from env args in docker run and so these variables will come from the packit deploy tool which makes sense

absternator commented on docker/Dockerfile:
```diff
@@ -17,3 +19,15 @@ RUN install2.r --error remotes &&  \
 COPY . /src
 RUN R CMD INSTALL --install-tests /src && rm -rf /src
+
+ENV ORDERLY_RUNNER_QUEUE_ID=orderly.runner
```
a more informative name just to be safe ahha eg. `orderly.runner.queueId`

M-Kusumgar replied:
gone with `orderly.runner.queue` for casing consistency

absternator reviewed (approved):
LGTM!!!! just some minor commnets.. 
- Maybe the scripts can be one level higher above test? in a scripts folder for consitency
- 

Also just a question... im gussiing in packet we just spin up the runner and worker right?

absternator reviewed (commented):
LGTM!!!! just some minor commnets.. 
- Maybe the scripts can be one level higher above test? in a scripts folder for consitency
- 

Also just a question... im gussiing in packet we just spin up the runner and worker right?

richfitz commented on docker/Dockerfile:
```diff
@@ -1,9 +1,11 @@
         libssl-dev \
         zlib1g-dev \
+        libhiredis-dev \
+        git \
```
can yo sort these alphabetically please?

M-Kusumgar replied:
done

richfitz commented on docker/Dockerfile:
```diff
@@ -17,3 +19,15 @@ RUN install2.r --error remotes &&  \
 RUN R CMD INSTALL --install-tests /src && rm -rf /src
+
+ENV ORDERLY_RUNNER_QUEUE_ID=orderly.runner
+ENV REDIS_CONTAINER_NAME=orderly.runner_redis
```
I don't think that this should be a default envvar in the image

M-Kusumgar replied:
removed

richfitz commented on docker/test/clear-test:
```diff
@@ -0,0 +1,22 @@
+set -x
+HERE=$(dirname $0)
+
+docker kill \
```
Use `docker rm -f` and you can avoid the repetition here

richfitz commented on docker/test/run-test:
```diff
@@ -0,0 +1,41 @@
+docker run --rm -d --pull=always \
+       --net=orderly.runner_network \
+       --name=orderly.runner_server \
+       --entrypoint="/usr/local/bin/orderly.runner.server" \
```
```suggestion
       --entrypoint="/usr/local/bin/orderly.runner.server" \
       --env=REDIS_CONTAINER_NAME=orderly.runner_redis \
```

etc

M-Kusumgar replied:
done

richfitz commented on docker/test/run-test:
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env bash
+set -ex
```
```suggestion
set -exu
```

Everywhere you want `-e` you usually want `-u`

M-Kusumgar replied:
done

M-Kusumgar committed:
review comments + general cleanup

M-Kusumgar commented:
I have addressed the comments but also done a bit of cleanup, all ive done that is extra is extract out certain variables you may want to change like container names, after approval I will also change it to pull image from main and merge it in so that we arent always pulling from this random branch. I will also remove triggers to the build and push workflow from this branch.

Also I remember this ticket that I had to do involving replacing `_` with `-` in docker container names, ive forgotten why I had to do it but to be on the safe side, the naming convention is now `orderly.runner-[container]` instead of `orderly.runner_[container]`

M-Kusumgar committed:
consistent variable

richfitz commented on R/main.R:
```diff
@@ -1,20 +1,47 @@
+main_worker <- function(args = commandArgs(TRUE)) {
+  dat <- parse_main_worker(args)
+
+  # assumes ORDERLY_RUNNER_QUEUE_ID is set
```
Does this error informatively if not set?

richfitz commented on R/queue.R:
```diff
@@ -25,9 +25,13 @@ Queue <- R6::R6Class("Queue", #nolint
       # Create queue
       self$controller <- rrq::rrq_controller(
-        queue_id %||% orderly_queue_id()
+        queue_id %||% orderly_queue_id(),
```
Suggest erroring here or orderly_queue_id() here if not set?  but we could explore that in another PR if you prefer?

M-Kusumgar replied:
i think we should error in the `main_worker` function because this is okay if we havent set the queue id, in fact we use that behaviour a lot in the tests, have made a ticket: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-5663/orderly.runner-worker-error-if-ORDERLYRUNNERQUEUEID-not-set

M-Kusumgar committed:
empty for ci

M-Kusumgar committed:
replace branchname with main for ci
